File: Julia/swapview.jl
#!/usr/bin/julia

using Printf

function filesizeKB(size::Int)
    left = float(size)
    unit = 1
    while left > 1100 && unit < 4
        left /= 1024
        unit += 1
    end
    @inbounds u = ("KMGT")[unit]
    @sprintf("%.1f%siB", left, u)::AbstractString
end

function getSwapFor(pid::AbstractString)
    s::Int = 0
    for m in eachmatch(r"Swap: *([0-9]*)", read("/proc/$pid/smaps", String))
        s += parse(Int, m.captures[1])
    end
    s
end

function getCmd(pid::AbstractString)
    comm::String = read("/proc/$pid/cmdline", String)
    !isempty(comm) && comm[end] == '\x00' && (comm = comm[1:end - 1])
    return replace(comm, "\x00" => " ")
end

function getSwap()
    ret = Any[]
    @inbounds for f in readdir("/proc")
        all(isdigit, f) && try
            s = getSwapFor(f)
            s > 0 && push!(ret, (f, s, getCmd(f)))
        catch
        end
    end
    sort!(ret, by=(x) -> x[2])
    return ret
end

function main()
    results = getSwap()
    @printf("%5s %9s %s\n", "PID", "SWAP", "COMMAND")
    totalsize = 0
    @inbounds for (pid, swap, comm) in results
        @printf("%5s %9s %s\n", pid, filesizeKB(swap), comm)
    end
    @printf("Total: %8s\n", isempty(results) ? "0B" :
            filesizeKB(sum((x) -> x[2], results)))
end

main()

File: PHP/swapview.php
#!/usr/bin/env php
<?php

function renderSize($size)
{
    $units = "KMGT";
    $left = $size;
    $unit = -1;

    while ($left > 1100 && $unit < 3) {
        $left /= 1024;
        $unit++;
    }

    if ($unit === -1) {
        return sprintf("%dB", $size);
    } else {
        if ($size < 0) {
            $left = -1 * $left;
        }

        return sprintf('%.1f%siB', $left, $units[$unit]);
    }
}


function getSwapFor($pid)
{
    $fallback = [$pid, 0, ''];
    $commFile = '/proc/' . $pid . '/cmdline';
    if(!($comm = @file_get_contents($commFile))) {
        return $fallback;
    }
    if ("\0" == substr($comm, -1)) {
        $comm = substr($comm, 0, strlen($comm) - 1);
    }

    $comm = str_replace("\0", " ", $comm);
    $smapsFile = '/proc/' . $pid . '/smaps';
    if (!($smaps = @file_get_contents($smapsFile))) {
        return $fallback;
    }

    $matchCount = preg_match_all('/\nSwap:\s+(\d+)/', $smaps, $matches);
    if (!$matchCount) {
        return $fallback;
    } else {
        $sum = array_sum($matches[1]);
        return array($pid, $sum * 1024, $comm);
    }
}


function getSwap()
{
    $ret = [];
    $dirs = scandir('/proc');
    foreach($dirs as $item)
    {
        $pid = intval($item);
        if($pid)
        {
            $swap = getSwapFor($pid);
            if($swap[1])
            {
                $ret[] = $swap;
            }
        }
    }
    array_multisort(array_column($ret, 1), SORT_ASC, $ret);
    return $ret;
}


$results = getSwap();
printf("%7s %9s %s", "PID", "SWAP", "COMMAND" . PHP_EOL);

$totalSize = 0;
foreach ($results as $result) {
    printf("%7s %9s %s" . PHP_EOL, $result[0], renderSize($result[1]), $result[2]);
    $totalSize += $result[1];
}

printf("Total: %8s" . PHP_EOL, renderSize($totalSize));
